File: web/src/utils/legacyFirestoreDedup.ts
export interface FirestorePulledDoc<T> {
  docId: string;
  record: T;
}

export interface DedupedPull<T> {
  records: T[];
  /** Physical Firestore doc ids to hard-delete (legacy numeric docs that are redundant). */
  legacyDocIdsToDelete: string[];
  /**
   * cloudIds of content-identical duplicate copies (distinct from the canonical copy).
   * These should be tombstoned in the cloud and removed from the local DB so every
   * device converges on a single copy.
   */
  duplicateCloudIds: string[];
  /**
   * Map of duplicate cloudId to canonical survivor cloudId. Used to reassign local linked rows
   * (like expenses) before the duplicate category is purged.
   */
  remapCloudIds: Record<string, string>;
}

interface InternalDoc<T> {
  docId: string;
  record: T;
  legacy: boolean;
}

export function isLegacyFirestoreDocId(docId: string): boolean {
  return docId.length > 0 && /^\d+$/.test(docId);
}

function isUuidShape(value: string): boolean {
  return value.length >= 32 && value.includes('-');
}

/** Prefer a modern (uuid) doc as the survivor, then the smallest cloudId for determinism. */
function chooseCanonical<T>(group: InternalDoc<T>[], cloudIdOf: (record: T) => string): InternalDoc<T> {
  return [...group].sort((a, b) => {
    if (a.legacy !== b.legacy) return a.legacy ? 1 : -1;
    return cloudIdOf(a.record).localeCompare(cloudIdOf(b.record));
  })[0];
}

function dedupeLegacyDocs<T>(
  docs: FirestorePulledDoc<T>[],
  fingerprint: (record: T) => string,
  cloudIdOf: (record: T) => string,
  collapseContentDuplicates: boolean,
  isRecovered: (record: T) => boolean = () => false,
): DedupedPull<T> {
  const internal: InternalDoc<T>[] = docs.map((doc) => ({
    docId: doc.docId,
    record: doc.record,
    legacy: isLegacyFirestoreDocId(doc.docId),
  }));

  const modern = internal.filter((doc) => !doc.legacy);
  const legacy = internal.filter((doc) => doc.legacy);
  const legacyDocIdsToDelete: string[] = [];
  const remapCloudIds: Record<string, string> = {};

  const modernByCloudId = new Map(modern.map((doc) => [cloudIdOf(doc.record), doc]));
  const modernByFingerprint = new Map<string, InternalDoc<T>[]>();
  for (const doc of modern) {
    const key = fingerprint(doc.record);
    const bucket = modernByFingerprint.get(key) ?? [];
    bucket.push(doc);
    modernByFingerprint.set(key, bucket);
  }

  // Phase 1: collapse legacy docs that are the same entity (cloudId pointer) or a
  // content duplicate of a modern doc.
  const keptLegacy: InternalDoc<T>[] = [];
  for (const leg of legacy) {
    const dataCloudId = cloudIdOf(leg.record);

    if (isUuidShape(dataCloudId) && dataCloudId !== leg.docId && modernByCloudId.has(dataCloudId)) {
      legacyDocIdsToDelete.push(leg.docId);
      remapCloudIds[leg.docId] = dataCloudId;
      continue;
    }

    const modernMatch = modernByFingerprint.get(fingerprint(leg.record))?.[0];
    if (modernMatch) {
      legacyDocIdsToDelete.push(leg.docId);
      remapCloudIds[leg.docId] = cloudIdOf(modernMatch.record);
      continue;
    }

    keptLegacy.push(leg);
  }

  const candidates = [...modern, ...keptLegacy];

  if (!collapseContentDuplicates) {
    return {
      records: candidates.map((doc) => doc.record),
      legacyDocIdsToDelete: [...new Set(legacyDocIdsToDelete)],
      duplicateCloudIds: [],
      remapCloudIds,
    };
  }

  // Phase 2: collapse content-identical duplicates among the survivors, keeping a single
  // canonical copy per fingerprint.
  const byFingerprint = new Map<string, InternalDoc<T>[]>();
  for (const doc of candidates) {
    const key = fingerprint(doc.record);
    const bucket = byFingerprint.get(key) ?? [];
    bucket.push(doc);
    byFingerprint.set(key, bucket);
  }

  const records: T[] = [];
  const duplicateCloudIds: string[] = [];
  const canonicalCloudIds = new Set<string>();

  for (const group of byFingerprint.values()) {
    const canonical = group.length === 1 ? group[0] : chooseCanonical(group, cloudIdOf);
    const canonicalCloudId = cloudIdOf(canonical.record);
    records.push(canonical.record);
    canonicalCloudIds.add(canonicalCloudId);

    if (group.length === 1) continue;

    for (const dup of group) {
      if (dup === canonical) continue;
      const dupCloudId = cloudIdOf(dup.record);
      remapCloudIds[dupCloudId] = canonicalCloudId;
      if (dupCloudId === canonicalCloudId) {
        // Same logical record, redundant physical doc.
        legacyDocIdsToDelete.push(dup.docId);
      } else {
        duplicateCloudIds.push(dupCloudId);
        if (dup.docId !== dupCloudId) legacyDocIdsToDelete.push(dup.docId);
      }
    }
  }

  // Phase 3: Special case for "Recovered" categories.
  // If a survivor is "Recovered", try to find a genuine category with same transaction type to merge into.
  const finalSurvivors = [...records];
  const recovered = finalSurvivors.filter(isRecovered);

  for (const rec of recovered) {
    const recCloudId = cloudIdOf(rec);
    const type = fingerprint(rec).split('|')[3] || '';
    // Find a standard category (not named "Recovered") with same transaction type
    const standardTarget = finalSurvivors.find((it) =>
      !isRecovered(it) && fingerprint(it).split('|')[3] === type
    );

    if (standardTarget) {
      const targetCloudId = cloudIdOf(standardTarget);
      duplicateCloudIds.push(recCloudId);
      remapCloudIds[recCloudId] = targetCloudId;
      const idx = finalSurvivors.indexOf(rec);
      if (idx > -1) finalSurvivors.splice(idx, 1);
    }
  }

  return {
    records: finalSurvivors,
    legacyDocIdsToDelete: [...new Set(legacyDocIdsToDelete)],
    duplicateCloudIds: [...new Set(duplicateCloudIds)].filter((cloudId) => !canonicalCloudIds.has(cloudId)),
    remapCloudIds,
  };
}

export function dedupeLegacyCategoryDocs<T extends {
  cloudId: string;
  name: string;
  iconName: string;
  colorInt: number;
  transactionType: string;
  sortOrder: number;
}>(docs: FirestorePulledDoc<T>[]): DedupedPull<T> {
  return dedupeLegacyDocs(
    docs,
    (record) => `${record.name}|${record.iconName}|${record.colorInt}|${record.transactionType}|${record.sortOrder}`,
    (record) => record.cloudId,
    true,
    (record) => record.name === 'Recovered',
  );
}

export function dedupeLegacyExpenseDocs<T extends {
  cloudId: string;
  amount: number;
  dateMillis: number;
  note: string;
  transactionType: string;
  receiptImagePath?: string | null;
}>(docs: FirestorePulledDoc<T>[]): DedupedPull<T> {
  return dedupeLegacyDocs(
    docs,
    (record) =>
      `${record.amount}|${record.dateMillis}|${record.note}|${record.transactionType}|${record.receiptImagePath ?? ''}`,
    (record) => record.cloudId,
    true,
  );
}

File: app/src/main/java/com/aus/ausgegeben/data/cloud/LegacyFirestoreDedup.kt
package com.aus.ausgegeben.data.cloud

data class FirestorePulledDoc<T>(
    val docId: String,
    val record: T,
)

data class DedupedPull<T>(
    val records: List<T>,
    /** Physical Firestore doc ids to hard-delete (legacy numeric docs that are redundant). */
    val legacyDocIdsToDelete: List<String>,
    /**
     * cloudIds of content-identical duplicate copies (distinct from the canonical copy). These
     * should be tombstoned in the cloud and removed from the local DB so every device converges
     * on a single copy.
     */
    val duplicateCloudIds: List<String> = emptyList(),
    /**
     * Map of duplicate cloudId to canonical survivor cloudId. Used to reassign local linked rows
     * (like expenses) before the duplicate category is purged.
     */
    val remapCloudIds: Map<String, String> = emptyMap()
)

fun isLegacyFirestoreDocId(docId: String): Boolean =
    docId.isNotEmpty() && docId.all { it.isDigit() }

private fun isUuidShape(value: String): Boolean =
    value.length >= 32 && value.contains('-')

fun dedupeLegacyCategoryDocs(docs: List<FirestorePulledDoc<CloudCategoryRecord>>): DedupedPull<CloudCategoryRecord> =
    dedupeLegacyDocs(
        docs = docs,
        fingerprint = { "${it.name}|${it.iconName}|${it.colorInt}|${it.transactionType}|${it.sortOrder}" },
        cloudIdOf = { it.cloudId },
        // COLLAPSE duplicates: if the user has redundant "Food" categories from different 
        // devices or legacy imports, merge them to prevent UI clutter.
        collapseContentDuplicates = true,
        isRecovered = { it.name == "Recovered" }
    )

fun dedupeLegacyExpenseDocs(docs: List<FirestorePulledDoc<CloudExpenseRecord>>): DedupedPull<CloudExpenseRecord> =
    dedupeLegacyDocs(
        docs = docs,
        fingerprint = {
            "${it.amount}|${it.dateMillis}|${it.note}|${it.transactionType}|${it.receiptImagePath.orEmpty()}"
        },
        cloudIdOf = { it.cloudId },
        collapseContentDuplicates = true,
    )

private data class InternalDoc<T>(
    val docId: String,
    val record: T,
    val legacy: Boolean,
)

/** Prefer a modern (uuid) doc as the survivor, then the smallest cloudId for determinism. */
private fun <T> chooseCanonical(group: List<InternalDoc<T>>, cloudIdOf: (T) -> String): InternalDoc<T> =
    group.sortedWith(compareBy({ it.legacy }, { cloudIdOf(it.record) })).first()

private fun <T> dedupeLegacyDocs(
    docs: List<FirestorePulledDoc<T>>,
    fingerprint: (T) -> String,
    cloudIdOf: (T) -> String,
    collapseContentDuplicates: Boolean,
    isRecovered: (T) -> Boolean = { false }
): DedupedPull<T> {
    val internal = docs.map { InternalDoc(it.docId, it.record, isLegacyFirestoreDocId(it.docId)) }
    val modern = internal.filter { !it.legacy }
    val legacy = internal.filter { it.legacy }
    val legacyDocIdsToDelete = mutableListOf<String>()
    val remapCloudIds = mutableMapOf<String, String>()

    val modernByCloudId = modern.associateBy { cloudIdOf(it.record) }
    val modernByFingerprint = modern.groupBy { fingerprint(it.record) }

    // Phase 1: collapse legacy docs that are the same entity (cloudId pointer) or a content
    // duplicate of a modern doc.
    val keptLegacy = mutableListOf<InternalDoc<T>>()
    for (leg in legacy) {
        val dataCloudId = cloudIdOf(leg.record)

        if (isUuidShape(dataCloudId) && dataCloudId != leg.docId && modernByCloudId.containsKey(dataCloudId)) {
            legacyDocIdsToDelete += leg.docId
            remapCloudIds[leg.docId] = dataCloudId
            continue
        }

        val modernMatch = modernByFingerprint[fingerprint(leg.record)]?.firstOrNull()
        if (modernMatch != null) {
            legacyDocIdsToDelete += leg.docId
            remapCloudIds[leg.docId] = cloudIdOf(modernMatch.record)
            continue
        }

        keptLegacy += leg
    }

    val candidates = modern + keptLegacy

    if (!collapseContentDuplicates) {
        return DedupedPull(
            records = candidates.map { it.record },
            legacyDocIdsToDelete = legacyDocIdsToDelete.distinct(),
            duplicateCloudIds = emptyList(),
            remapCloudIds = remapCloudIds
        )
    }

    // Phase 2: collapse content-identical duplicates among the survivors, keeping a single
    // canonical copy per fingerprint.
    val byFingerprint = candidates.groupBy { fingerprint(it.record) }
    val records = mutableListOf<T>()
    val duplicateCloudIds = mutableListOf<String>()
    val canonicalCloudIds = mutableSetOf<String>()

    for ((_, group) in byFingerprint) {
        val canonical = if (group.size == 1) group[0] else chooseCanonical(group, cloudIdOf)
        val canonicalCloudId = cloudIdOf(canonical.record)
        records += canonical.record
        canonicalCloudIds += canonicalCloudId

        if (group.size == 1) continue

        for (dup in group) {
            if (dup === canonical) continue
            val dupCloudId = cloudIdOf(dup.record)
            remapCloudIds[dupCloudId] = canonicalCloudId
            if (dupCloudId == canonicalCloudId) {
                legacyDocIdsToDelete += dup.docId
            } else {
                duplicateCloudIds += dupCloudId
                if (dup.docId != dupCloudId) legacyDocIdsToDelete += dup.docId
            }
        }
    }
    
    // Phase 3: Special case for "Recovered" categories. 
    // If a survivor is "Recovered", try to find a genuine category with same transaction type to merge into.
    val finalSurvivors = records.toMutableList()
    val recovered = finalSurvivors.filter(isRecovered)
    
    for (rec in recovered) {
        val recCloudId = cloudIdOf(rec)
        val type = fingerprint(rec).split('|').getOrNull(3) ?: ""
        // Find a standard category (not named "Recovered") with same transaction type
        val standardTarget = finalSurvivors.firstOrNull { 
            !isRecovered(it) && fingerprint(it).split('|').getOrNull(3) == type 
        }
        
        if (standardTarget != null) {
            val targetCloudId = cloudIdOf(standardTarget)
            duplicateCloudIds += recCloudId
            remapCloudIds[recCloudId] = targetCloudId
            finalSurvivors.remove(rec)
        }
    }

    return DedupedPull(
        records = finalSurvivors,
        legacyDocIdsToDelete = legacyDocIdsToDelete.distinct(),
        duplicateCloudIds = duplicateCloudIds.distinct().filterNot { it in canonicalCloudIds },
        remapCloudIds = remapCloudIds
    )
}
